RFDs/0001-job-worker-service.md

```diff
@@ -0,0 +1,468 @@
+```
+
+### 2. API
+
+#### gRPC API Specification
+
+```protobuf
+syntax = "proto3";
+
+package jobworker;
+
+import "google/protobuf/timestamp.proto";
+
+// JobWorker service provides methods to manage and monitor jobs.
+service JobWorker {
+  // Starts a new job with the given command and resource limits.
+  rpc StartJob(StartJobRequest) returns (StartJobResponse) {}
+  
+  // Stops a running job identified by its ID.
+  rpc StopJob(JobId) returns (StopJobResponse) {}
+  
+  // Retrieves the current status of a job.
+  rpc GetJobStatus(JobId) returns (stream JobStatusResponse) {}
+  
+  // Streams the output of a running job.
+  rpc GetJobOutput(JobId) returns (stream JobOutputResponse) {}
+
+  // Returns the list of all the jobs for the user
+  rpc ListJobs() returns (ListJobsResponse) {}
+}
+
+// StartJobRequest contains the information needed to start a new job.
+message StartJobRequest {
+  // The command to be executed.
+  string command = 1;
+  
+  // The arguments for the command.
+  repeated string args = 2;
+  
+  // The resource limits for the job.
+  ResourceLimits resource_limits = 3;
+}
+
+message JobId {
+  // The UUID of the job.
+  string id = 1;
+}
+
+// StartJobResponse contains the result of a job start request.
+message StartJobResponse {
+  // The UUID of the started job.
+  string id = 1;
+  
+  // The initial status of the job.
+  JobStatus status = 2;
+}
+
+// StopJobResponse contains the result of a job stop request.
+message StopJobResponse {
+  // The UUID of the stopped job.
+  string id = 1;
+  
+  // The final status of the job.
+  JobStatus status = 2;
+  
+  // The time when the job was started.
+  google.protobuf.Timestamp start_time = 3;
+  
+  // The time when the job was stopped.
+  google.protobuf.Timestamp stop_time = 4;
+}
+
+// JobStatusResponse contains the current status and resource usage of a job.
+message JobStatusResponse {
+  // The current status of the job.
+  JobStatus status = 1;
+  
+  // The current resource usage of the job.
+  ResourceUsage resource_usage = 2;
+}
+
+// JobOutputResponse contains a chunk of output data from a job.
+message JobOutputResponse {
+  // A chunk of output data from the job.
+  bytes data = 1;
+}
+
+// ResourceLimits defines the resource constraints for a job.
+message ResourceLimits {
+  // The number of CPU shares allocated to the job.
+  uint64 cpu_shares = 1;
+  
+  // The memory limit in bytes.
+  uint64 memory_limit_bytes = 2;
+  
+  // The I/O weight for prioritizing disk I/O.
+  uint64 io_weight = 3;
+}
+
+// ResourceUsage represents the current resource consumption of a job.
+message ResourceUsage {
+  // The current CPU usage.
+  uint64 cpu_usage = 1;
+  
+  // The current memory usage in bytes.
+  uint64 memory_usage = 2;
+  
+  // The current I/O usage.
+  uint64 io_usage = 3;
+}
+
+message ListJobsRequest {
+  // List of user ids to retrieve running jobs for
+  // Will return all if empty
+  repeated string user_ids = 1;
+}
+
+// List of jobs by user
```

Comment:
Listing jobs is not a challenge requirement

Reply:
True. I had added it as a nice to have. Getting rid of it in revised design.